go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+		return nil, fmt.Errorf("error getting table %s from commit: %v", tableName, err)
+	}
+	if t == nil {
+		return nil, fmt.Errorf("no table named %s found in commit", tableName)
+	}
+
+	schema, err := t.GetSchema(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema from table %s: %v", tableName, err)
+	}
+
+	return schema, nil
+}
+
+func pkColNamesFromCommit(ctx context.Context, c *doltdb.Commit, tableName string) ([]string, error) {
+	schema, err := schemaFromCommit(ctx, c, tableName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema for commit: %v", err)
+	}
+	return schema.GetPKCols().GetColumnNames(), nil
+}
+
+// rowChanged returns true if the row identified by `rowPK` in `tableName` changed between commits `old` and `new`
+// It's a bit inefficient when used repeatedly on adjacent commits such as in blame; each iteration it will
+// retrieve a row that was already retrieved on the previous iteration (i.e. last iteration's "old" row
+// is this iteration's "new" row).
+func rowChanged(ctx context.Context, old, new *doltdb.Commit, tableName string, rowPK types.Value) (bool, error) {
+	oldHash, newHash, err := getCommitHashes(old, new)
+	if err != nil {
+		return false, err
+	}
+
+	oldTable, err := maybeTableFromCommit(ctx, old, tableName)
+	if err != nil {
+		return false, fmt.Errorf("error getting table from old commit %s: %v", oldHash, err)
+	}
+	newTable, err := maybeTableFromCommit(ctx, new, tableName)
+	if err != nil {
+		return false, fmt.Errorf("error getting table from new commit %s: %v", newHash, err)
+	}
+
+	// if the table is in the old commit but not the new one...something's wrong. bail!
+	if oldTable != nil && newTable == nil {
+		return false, fmt.Errorf("expected to find table with name %v in new commit %s, but didn't", tableName, newHash)
+	}
+	// if the table is in the new commit but not the old one, it must be new; return true
+	if newTable != nil && oldTable == nil {
+		return true, nil
+	}
+
+	oldRow, err := maybeRowFromTable(ctx, oldTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in old commit %s: %v", tableName, oldHash, err)
+	}
+	newRow, err := maybeRowFromTable(ctx, newTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in new commit %s: %v", tableName, newHash, err)
+	}
+
+	// if the row is in the old table but not the new one...something's wrong. bail!
+	if oldRow != nil && newRow == nil {
+		return false, fmt.Errorf("expected to find row with PK %v in table %s in new commit %s, but didn't", rowPK, tableName, newHash)
+	}
+	// if the row is in the new table but not the old one, it must be new; return true
+	if newRow != nil && oldRow == nil {
+		return true, nil
+	}
+
+	oldSchema, err := oldTable.GetSchema(ctx)
+	if err != nil {
+		return false, fmt.Errorf("error getting schema from %s in old commit %s: %v", tableName, oldHash, err)
+	}
+
+	return !row.AreEqual(*oldRow, *newRow, oldSchema), nil
```

Comment:
Does `row.AreEqual` return false if a column was added in `newTable`'s schema, `newRow` was populated with a non-nil value for that column and then we called `row.AreEqual(*oldRow, *newRow, oldSchema)`?

Reply:
Yeah that's a good question. The existing tests don't consider this case but I'm trying to add it now.